File: backend/lightning.py
"""Lightning from NASA, by way of the GOES Lightning Mappers.

Every GOES weather satellite carries a GLM -- a camera staring at the whole
disc of the earth, 500 times a second, watching for the flash of a lightning
stroke through the cloud tops. NASA republishes it through GIBS as map tiles,
the same service this app already uses for the daily cloud mosaic, so it needs
no account, no key and no parsing: it is a tile layer like any other.

What it does not do is cover the world. The GOES satellites sit over the
Americas -- East at 75 degrees west, West at 137 -- and a geostationary camera
sees rather less than a hemisphere. Europe, Africa and Asia are off the edge of
both, and no amount of waiting will put lightning there. That is a property of
where the satellites are, not a fault, and the panel says so plainly, because
a blank map that means "nothing is happening here" and a blank map that means
"nothing can be seen from here" look exactly alike.

The layer identifiers are not written down here. GIBS publishes its own
catalogue, so this asks for it and takes whatever GLM layers it actually
offers, with the time extents it actually has. Hard-coding a name is how you
ship a layer that quietly returns nothing the day it is renamed -- and how the
last version of this file failed, betting on a service whose protocol could
not be checked from here.
"""

from __future__ import annotations

import threading
import time
import xml.etree.ElementTree as ET
from typing import Any

import requests

from . import config

ATTRIBUTION = "NASA GIBS · GOES GLM"

# The sentence the panel shows under an empty map. Written here beside the
# reason for it rather than in the page, so it cannot drift away from the fact.
COVERAGE = (
    "GOES watches the Americas and the oceans either side of them. Europe, "
    "Africa and Asia are over the horizon from both satellites, so they stay "
    "blank however active the weather is."
)

# GIBS's own catalogue of what it serves, in the projection the map uses.
CAPABILITIES = ("https://gibs.earthdata.nasa.gov/wmts/epsg3857/best"
                "/1.0.0/WMTSCapabilities.xml")

# The tile template every GIBS layer follows. The front end fills this in.
TILES = ("https://gibs.earthdata.nasa.gov/wmts/epsg3857/best"
         "/{layer}/default/{time}/{matrix}/{z}/{y}/{x}.{fmt}")

# What a GLM layer is called, loosely enough to survive renaming.
#
# The first version wanted "glm" *and* "flash" in the identifier, which found
# nothing at all: too clever by half, and a filter that has to be right about
# two words is twice as likely to be wrong about one. Either word will do now,
# in the identifier or the human-readable title, and "lightning" as well --
# whatever NASA calls it, one of these three is in the name.
WANTED = ("glm", "lightning")

# A wider net, used only to explain a miss. If nothing matches above, these are
# the layers worth showing to whoever has to work out why -- otherwise the
# answer is "no layers" with nothing to go on, which is where the last attempt
# left off.
NEARBY = ("glm", "lightning", "flash", "goes")

# The greyscale cloud-top picture the flashes are usually drawn over. Band 13
# is the clean infrared window -- what the satellite sees of cloud temperature,
# day or night, which is why a storm looks like a storm in it. GeoColor is the
# prettier composite and goes dark at night, so the infrared is preferred.
BACKDROP = ("band13", "clean_infrared", "cleanir", "geocolor", "infrared")

# The catalogue is a couple of megabytes and changes about never. Fetched once
# and held for the day.
CACHE_SECONDS = 6 * 3600

_NS = {
    "wmts": "http://www.opengis.net/wmts/1.0",
    "ows": "http://www.opengis.net/ows/1.1",
}

_lock = threading.Lock()
_cached: dict[str, Any] | None = None
_cached_at = 0.0


class LightningError(RuntimeError):
    pass


def _entry(layer: ET.Element, ident: str, title: str, low: str) -> dict[str, Any]:
    """One layer, in the shape the front end needs to build a tile URL."""
    matrix = layer.findtext(
        f"{{{_NS['wmts']}}}TileMatrixSetLink/{{{_NS['wmts']}}}TileMatrixSet")
    fmt = layer.findtext(f"{{{_NS['wmts']}}}Format") or "image/png"

    default = None
    values: list[str] = []
    for dim in layer.findall(f"{{{_NS['wmts']}}}Dimension"):
        if (dim.findtext(f"{{{_NS['ows']}}}Identifier") or "").lower() != "time":
            continue
        default = dim.findtext(f"{{{_NS['wmts']}}}Default")
        values = [v.text for v in dim.findall(f"{{{_NS['wmts']}}}Value") if v.text]

    return {
        "id": ident,
        "title": title,
        "matrix": matrix,
        "format": (fmt.rsplit("/", 1)[-1] or "png").replace("jpeg", "jpg"),
        "time_default": default,
        "time_values": values,
        # East and West see different halves of the Americas, and which one a
        # viewer wants depends on where they are looking.
        "satellite": "west" if "west" in low else "east" if "east" in low else None,
    }


def parse_layers(xml: str) -> dict[str, list[dict[str, Any]]]:
    """Pull the GOES layers out of a WMTS capabilities document.

    Two sorts, because the picture people mean by "lightning from GOES" is two
    layers: the flashes themselves, and the greyscale infrared cloud tops they
    are drawn over. On their own the flashes are dots in a void.
    """
    try:
        root = ET.fromstring(xml)
    except ET.ParseError as exc:
        raise LightningError(f"GIBS sent a catalogue that would not parse: {exc}") from exc

    flashes: list[dict[str, Any]] = []
    cloud: list[dict[str, Any]] = []
    for layer in root.iter(f"{{{_NS['wmts']}}}Layer"):
        ident = layer.findtext(f"{{{_NS['ows']}}}Identifier") or ""
        title = layer.findtext(f"{{{_NS['ows']}}}Title") or ident
        # Both, because GIBS is not consistent about which one carries the
        # instrument name.
        low = f"{ident} {title}".lower()

        if any(word in low for word in WANTED):
            flashes.append(_entry(layer, ident, title, low))
        elif "goes" in low and any(word in low for word in BACKDROP):
            cloud.append(_entry(layer, ident, title, low))

    flashes.sort(key=lambda entry: entry["id"])
    cloud.sort(key=lambda entry: entry["id"])
    return {"lightning": flashes, "imagery": cloud}


def near_misses(xml: str, limit: int = 12) -> list[str]:
    """Layer names worth looking at when nothing matched.

    A miss that says only "no layers" is a dead end -- it cannot tell you
    whether NASA has stopped publishing, renamed the product, or whether the
    filter is simply wrong. This lists what is actually there under any related
    word, so the next step is reading a list rather than guessing again.
    """
    try:
        root = ET.fromstring(xml)
    except ET.ParseError:
        return []
    seen: list[str] = []
    for layer in root.iter(f"{{{_NS['wmts']}}}Layer"):
        ident = layer.findtext(f"{{{_NS['ows']}}}Identifier") or ""
        if any(word in ident.lower() for word in NEARBY) and ident not in seen:
            seen.append(ident)
        if len(seen) >= limit:
            break
    return seen


def _fetch() -> tuple[list[dict[str, Any]], list[str], int]:
    try:
        resp = requests.get(CAPABILITIES, timeout=30,
                            headers={"User-Agent": config.USER_AGENT})
    except requests.RequestException as exc:
        raise LightningError(f"NASA GIBS could not be reached: {exc}") from exc
    if not resp.ok:
        raise LightningError(f"NASA GIBS answered {resp.status_code}")
    found = parse_layers(resp.text)
    total = resp.text.count("</Layer>")
    misses = [] if found["lightning"] else near_misses(resp.text)
    return found, misses, total


def layers(refresh: bool = False) -> dict[str, Any]:
    """The GLM layers GIBS is currently serving, with the tile template."""
    global _cached, _cached_at
    with _lock:
        fresh = _cached is not None and time.time() - _cached_at < CACHE_SECONDS
        if fresh and not refresh:
            return dict(_cached)

    found, nearby, total = _fetch()
    answer = {
        "layers": found["lightning"],
        "imagery": found["imagery"],
        "template": TILES,
        "attribution": ATTRIBUTION,
        "coverage": COVERAGE,
        # Only populated when nothing matched, and then it is the whole point
        # of the response.
        "nearby": nearby,
        "catalogue_size": total,
    }
    with _lock:
        _cached = answer
        _cached_at = time.time()
    return dict(answer)


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def demo() -> dict[str, Any]:
    """What the layer picker gets in the demo build.

    The same shape as the real answer, with the identifiers GIBS used when this
    was written. Demo mode has no network, and a picker with nothing in it
    would make the layer look broken rather than offline.
    """
    return {
        "layers": [
            {"id": "GOES-East_GLM_Flash_Extent_Density",
             "title": "GOES-East GLM Flash Extent Density (demo)",
             "matrix": "GoogleMapsCompatible_Level7", "format": "png",
             "time_default": "default", "time_values": [], "satellite": "east"},
            {"id": "GOES-West_GLM_Flash_Extent_Density",
             "title": "GOES-West GLM Flash Extent Density (demo)",
             "matrix": "GoogleMapsCompatible_Level7", "format": "png",
             "time_default": "default", "time_values": [], "satellite": "west"},
        ],
        "imagery": [
            {"id": "GOES-East_ABI_Band13_Clean_Infrared",
             "title": "GOES-East infrared cloud tops (demo)",
             "matrix": "GoogleMapsCompatible_Level7", "format": "png",
             "time_default": "default", "time_values": [], "satellite": "east"},
            {"id": "GOES-West_ABI_Band13_Clean_Infrared",
             "title": "GOES-West infrared cloud tops (demo)",
             "matrix": "GoogleMapsCompatible_Level7", "format": "png",
             "time_default": "default", "time_values": [], "satellite": "west"},
        ],
        "template": TILES,
        "attribution": "synthetic",
        "coverage": COVERAGE,
        "nearby": [],
        "catalogue_size": 4,
    }

File: frontend/js/lightning.js
// Lightning, from NASA's GOES Lightning Mappers.
//
// Every GOES satellite carries a camera watching the whole disc of the earth
// five hundred times a second for the flash of a stroke through the cloud
// tops. NASA republishes it through GIBS as map tiles -- the same service the
// daily cloud mosaic already comes from -- so this is a tile layer and nothing
// more: no account, no key, no stream to hold open.
//
// The layer names are asked for rather than written down. GIBS publishes its
// own catalogue and the backend reads the GLM entries out of it, so a rename
// changes what appears in the picker instead of quietly serving nothing.
//
// The thing to be clear about is where it can see. GOES sits over the
// Americas, and a geostationary camera sees less than a hemisphere: Europe,
// Africa and Asia are over its horizon and stay blank whatever the weather is
// doing there. An empty map that means "quiet" and one that means "not
// watched from here" look identical, so the panel always says which.

import { api } from './api.js';
import { $, el } from './ui.js';

// Roughly where each satellite can see anything at all: a geostationary
// camera at this height is useful to about 65 degrees from the point beneath
// it, and hopeless past that. Used only to tell someone why their map is
// empty, never to hide the layer.
const SEEN_FROM = {
  east: { lon: -75.2, label: 'GOES-East' },
  west: { lon: -137.2, label: 'GOES-West' },
};
const REACH_DEGREES = 65;

// How often to ask GIBS for a newer frame. The product is published every few
// minutes; this is the cheapest cadence that keeps up with it.
const REFRESH_MS = 120000;

let map = null;
let layer = null;
let enabled = false;
let catalogue = null;
let chosen = null;
let opacity = 0.85;
let timer = null;
let problem = '';
// Until somebody picks a satellite themselves, the map picks the one looking
// at whatever is on screen. After that it stays put: an automatic choice that
// overrides a deliberate one is worse than no automatic choice at all.
let manual = false;

export function initLightning(leafletMap) {
  map = leafletMap;
  // Above the radar: rain is the context, the strike is the event, and the
  // point of having both is to see one inside the other.
  map.createPane('lightning').style.zIndex = 465;
  map.getPane('lightning').style.pointerEvents = 'none';
  buildDock();
  map.on('moveend', () => {
    if (!enabled) return;
    // Panning from Florida to California is a change of satellite, not just of
    // view -- so the choice is remade, and only redrawn if it actually moved.
    if (!manual) {
      const next = pick();
      if (next && next !== chosen) { chosen = next; buildDock(); show(); return; }
    }
    paintDock();
  });
}

// ── The layer ──────────────────────────────────────────────────

/** The tile URL for one catalogue entry, with the placeholders filled in. */
function urlFor(entry) {
  return catalogue.template
    .replace('{layer}', entry.id)
    // GIBS wants a time or the word default; a layer with no time dimension
    // has nothing to substitute and takes the latter.
    .replace('{time}', entry.time_default ?? 'default')
    .replace('{matrix}', entry.matrix ?? 'GoogleMapsCompatible_Level7')
    // Leaflet fills in {z}, {y} and {x} itself and throws on anything else it
    // is handed, so the extension has to be substituted here rather than left
    // in the template for it to trip over.
    .replace('{fmt}', entry.format ?? 'png');
}

function show() {
  layer?.remove();
  layer = null;
  if (!enabled || !chosen) return;

  let missing = 0;
  layer = L.tileLayer(urlFor(chosen), {
    pane: 'lightning',
    opacity,
    // GIBS stops at the zoom the product is gridded to; past that the tiles
    // are stretched rather than withheld, so the layer stays on screen when
    // you zoom into a storm instead of vanishing.
    maxNativeZoom: 7,
    maxZoom: 19,
    className: 'bolt-tiles',
    attribution: catalogue.attribution,
  });
  // A missing tile over the ocean is ordinary; every tile missing means the
  // layer name is wrong or the product has moved, and that is worth saying
  // rather than showing an empty map and letting it read as calm weather.
  layer.on('tileerror', () => {
    missing += 1;
    if (missing === 8) { problem = `${chosen.id} is not answering.`; paintDock(); }
  });
  layer.on('tileload', () => {
    if (!problem) return;
    problem = '';
    paintDock();
  });
  layer.addTo(map);
}

async function load() {
  try {
    catalogue = await api.lightning();
    // A miss that says only "none" leaves nowhere to go. If GIBS answered with
    // a catalogue but nothing in it looked like lightning, say how big the
    // catalogue was and name what came closest -- that is the difference
    // between a dead end and a next step.
    problem = catalogue.layers.length ? '' : noneFound(catalogue);
    chosen = pick();
    buildDock();
    show();
  } catch (err) {
    problem = err.message;
    paintDock();
  }
}

/** What to say when GIBS answered but had no lightning layer in it. */
function noneFound(got) {
  const size = got.catalogue_size ? ` of ${got.catalogue_size}` : '';
  if (got.nearby?.length) {
    return `No lightning layer${size} at NASA GIBS. Closest names there: `
      + `${got.nearby.slice(0, 5).join(', ')}.`;
  }
  return `NASA GIBS is serving no lightning layer${size} in this projection.`;
}

/** Whichever satellite is looking at the middle of the map. */
function pick() {
  if (!catalogue?.layers?.length) return null;
  const centre = map.getCenter().lng;
  const distance = (entry) => {
    const seen = SEEN_FROM[entry.satellite];
    if (!seen) return 999;
    return Math.abs(((centre - seen.lon + 540) % 360) - 180);
  };
  return [...catalogue.layers].sort((a, b) => distance(a) - distance(b))[0];
}

/** How far the map's centre is from the nearest satellite's view. */
function reach() {
  const seen = SEEN_FROM[chosen?.satellite];
  if (!seen) return null;
  const centre = map.getCenter().lng;
  const away = Math.abs(((centre - seen.lon + 540) % 360) - 180);
  return { away, label: seen.label, inside: away <= REACH_DEGREES };
}

// ── The panel ──────────────────────────────────────────────────

function buildDock() {
  const dock = $('#lightningDock');
  if (!dock) return;
  dock.innerHTML = '';
  dock.append(
    el('button', { class: 'bolt-toggle', id: 'boltToggle', onclick: toggle },
      el('span', { class: 'bolt-mark' }, '⚡'), 'Lightning'),
    el('div', { class: 'bolt-body', id: 'boltBody', hidden: !enabled },
      catalogue?.layers?.length > 1
        ? el('div', { class: 'bolt-windows' },
          ...catalogue.layers.map((entry) => el('button', {
            class: `bolt-window${entry === chosen ? ' is-on' : ''}`,
            dataset: { layer: entry.id },
            title: entry.title,
            onclick: () => { chosen = entry; manual = true; buildDock(); show(); },
          }, SEEN_FROM[entry.satellite]?.label ?? entry.id.slice(0, 12))))
        : null,
      el('div', { class: 'bolt-count', id: 'boltCount' }, 'Loading…'),
      el('div', { class: 'bolt-key' },
        el('span', { class: 'bolt-key-row' },
          el('i', { style: 'background:#2b1d5e' }), 'few'),
        el('span', { class: 'bolt-key-row' },
          el('i', { style: 'background:#c86bff' }), 'many flashes')),
      el('div', { class: 'bolt-note', id: 'boltNote' }, '')));
  paintDock();
}

function toggle() {
  enabled = !enabled;
  $('#boltToggle').classList.toggle('is-on', enabled);
  $('#boltBody').hidden = !enabled;
  if (enabled) {
    if (catalogue) { chosen ??= pick(); show(); } else load();
    // Re-made rather than refreshed: the time is baked into the URL, so a
    // newer frame is a different layer.
    timer = setInterval(() => {
      if (!enabled) return;
      if (!manual) chosen = pick();
      load();
    }, REFRESH_MS);
  } else {
    clearInterval(timer);
    timer = null;
    layer?.remove();
    layer = null;
  }
  paintDock();
}

function paintDock() {
  const count = $('#boltCount');
  const text = $('#boltNote');
  if (!count || !text || !enabled) return;

  const view = reach();
  count.textContent = chosen
    ? (view && !view.inside
      ? `Nothing to see here — ${view.label} cannot look this far`
      : `${chosen.title.replace(/ \(demo\)$/, '')}`)
    : 'No lightning layer available';

  const lines = [];
  if (view && !view.inside) {
    lines.push(catalogue?.coverage ?? '');
  } else if (catalogue?.attribution === 'synthetic') {
    lines.push('Demo mode: the tiles are stand-ins.');
  } else if (chosen) {
    lines.push(`${catalogue.attribution}. Flash extent density — how many `
      + 'flashes touched each cell, not single strokes.');
  }
  if (problem) lines.push(problem);
  text.textContent = lines.filter(Boolean).join(' ');
}
